### DISCUSSION TITLE
Support for IGMPv3

### DISCUSSION THREAD

**Discussion by janowi** [troglobit/smcroute#203] [Q&A] [answered]
Would it be possible to add support for IGMPv3 so during the join we can already specify source from which we want data?

**Comment by troglobit** [comment]
Not sure I understand what you mean. Do you mean querier functionality?  Because you can already join SSM.

**Reply by janowi** [reply]
ok maybe I'm not aware then, if I would like to receive traffic from source 10.0.0.1 from group like 224.0.2.10 what kind of command I would need to execute? because smcroutectl eth0 224.0.2.10 sends IGMPv2 so I have no control on client from what source I will get the traffic. Also since host doesn't provide the source we need to use RP and then switchover to SSM

**Reply by troglobit** [reply]
I believe your questions are covered by the documentation I've written. Please have a look at

https://man.troglobit.com/man8/smcroutectl.8.html

and

https://github.com/troglobit/smcroute#usage

**Reply by janowi** [reply]
It's not like I made this post because I didn't read the readme or man, I made this post because I don't see any possibility to specify source from which I want traffic for multicast group.

**Reply by troglobit** [reply, marked as answer]
In the documentation you can see the syntax 'join IFNAME SOURCE GROUP', it works like this:

    $ smcroutectl join eth1 10.0.0.11 225.1.2.3

The join command is translated to an API in Linux, defined in [RFC 3678](https://www.rfc-editor.org/rfc/rfc3678), which uses `struct group_source_req` with the socket option `MCAST_JOIN_SOURCE_GROUP`. This in turn is translated into an IGMPv3 join message on the wire.  Here's a screenshot from a wireshark I just ran:

![igmp-screenshot](https://github.com/troglobit/smcroute/assets/183517/1a9584a4-0826-4c22-8209-9ef817896310)

Provided the IP stack has not detected an IGMPv2 message on the LAN, which will have caused it to fall back to IGMPv2.  This might very well have happened in your case, but that is outside the scope of SMCRoute (or any other multicast routing application), since the standard clearly defines what happens if a network is not fully IGMPv3 compliant.

I hope this clears up any confusion you might have had.
